### DISCUSSION TITLE
How to set gravity in PeleC?

### DISCUSSION THREAD

**Discussion by Alex-jian522** [AMReX-Combustion/PeleC#753] [Q&A] [answered]
Hi all,

I want to simulate a spherical flame affected by gravity using peleC. I have set all the boundary conditions to be "FOExtrap". But I can not change the gravity?
Is there a command to change gravity in PeleC or other alternatives?

Thank you very much!!

**Comment by marchdf** [comment, marked as answer]
We removed gravity since that was a legacy option from an astrophysics code. However, you can get the same functionality by usoing the following file: https://github.com/AMReX-Combustion/PeleC/blob/development/Source/External.cpp. You can use that as a template to add your own forcing term or use 
```
pelec.add_ext_src = 1
pelec.external_forcing = 1.0 0.0 0.0
```
and you can see how that code will treat those numbers. Hopefully this helps!

**Reply by Alex-jian522** [reply]
Thank you very much for your prompt reply.
When simulating a 2D spherical flame (without gravity forces yet), I set all four boundary conditions to "FOExtrap" (outflow). However, the velocity result cloud diagram shows that such boundary setting is unreasonable. I would like to ask if there are more suitable boundary condition settings?
Also, I had a similar error using outflow boundary conditions in peleLMeX.

![image](https://github.com/AMReX-Combustion/PeleC/assets/72745362/24c6f9ff-3799-4e59-8e51-c8128b689407)

**Reply by marchdf** [reply]
Hi, I am not sure what this case is and what "reasonable" looks like so I am not sure how to guide a discussion about BC. You could specify explicit BC using bcnormal and there are a lot of examples in the code that do that (e.g. implement a Blazek BC). Let us know if that doesn't suit your case.

**Reply by Alex-jian522** [reply]
This case is a 2-dimensional spherical premixed flame. I think a reasonable velocity distribution should not have low velocity areas (velocity magnitude is almost 0) at the four corners.
Thank you again for your patient answer.

**Reply by marchdf** [reply]
Can you post your input file? And maybe your case files?

**Reply by Alex-jian522** [reply]
[input.txt](https://github.com/AMReX-Combustion/PeleC/files/14119185/input.txt)
[prob.txt](https://github.com/AMReX-Combustion/PeleC/files/14119186/prob.txt)
Of course, the input file and case file are in the attachment. I tried changing the exit boundary to Blazek BC, but the results diverge. It may be because I'm not very familiar with pelec's boundary settings. The file suffix was changed to txt for easier uploading.

**Reply by marchdf** [reply]
Hi, can you post all the files such that we can reproduce what you are doing?

**Reply by baperry2** [reply]
It may be easiest to do this by creating a fork of PeleC and pushing branch with the relevant files to reproduce your case, and then linking to that branch here.

**Reply by Alex-jian522** [reply]
I'm very sorry that I didn't see this reply in time. I put all the files in this case into a compressed package:
"NH3-19SP" is the chemical mechanism we used.
"xlookup.dat" is the 1D premixed flame data used to initilize the flame.
All the prob files are modified from the case "HotBubble".
Thank you very much for your help. If you have any questions, please feel free to ask and I will reply promptly.
[sphere_flame.zip](https://github.com/AMReX-Combustion/PeleC/files/14174418/sphere_flame.zip)

**Reply by marchdf** [reply]
I ran your case and don't see anything "wrong" a-priori. You have a cylindrically symmetric solution inside a square domain. It feels inevitable that things are going to lose cylindrical symmetry at the domain BCs. Unless I am totally missing something. If you turn off diffusion, momentum is near machine precision everywhere. If you are worried about boundary effects, you can make your domain larger and use AMR.

**Reply by baperry2** [reply]
Yeah I agree with @marchdf. Now that I understand the case, this is a fundamental issue with applying FOextrap BCs (or outflow BCs in PeleLMeX) while trying to simulate something cylindrically symmetric. These boundary conditions state that du/dx = 0 on the x-normal boundaries and dv/dy = 0 on y-normal boundaries. But for a cylindrically symmetric flow in the far field, velocity decays as 1/r, so the BCs are incompatible with your desired solution. The best solution is to use a large domain so the boundaries are far away from the region of interest and the distortions in the flow field that result don't affect the region you care about. With AMR, this should not add much cost.

**Reply by Alex-jian522** [reply]
Thank you very much for your replies, I think I will adopt a larger computational domain and use symmetry boundaries and AMR to reduce the computational cost. And I will use PeleLMeX since it is convinient to set gravity in PeleLMeX. Thank you again!!

**Reply by baperry2** [reply]
Note that symmetry BCs (equivalent to slip walls) would change your problem to a confined domain rather than an open domain, which fundamentally changes the nature of the problem. No need to change the BC unless you intend that, a larger domain should be sufficient. I do think PeleLMeX is more appropriate as this is likely a low speed flow.

**Reply by Alex-jian522** [reply]
Thank you very much, I understand now.
